docs/decisions/0004-explicitly-declare-enterprise-for-role-assignments.rst

```diff
@@ -0,0 +1,73 @@
+Enterprise role assignments will record a customer UUID
+-------------------------------------------------------
+
+Status
+======
+
+Proposed - January 2021
+
+Context
+=======
+
+The ``SystemWideEnterpriseUserRoleAssignment`` model is used to grant role-based permissions
+to enterprise users.  These permissions generally control the users' access to data
+and ability to perform certain functions within an enterprise customer.  This assignment model does `not` currently
+refer to any specific customer - the "context" to which a role is applied is inferred from
+a user (an ``auth.User`` record) having been "linked" to a customer via the ``EnterpriseCustomerUser`` model.
+
+However, a single ``auth.User`` may be linked to multiple ``EnterpriseCustomers``, and if that user
+is assigned a role (such as ``enterprise_admin``), it is inferred that the role assignment applies
+to **every** customer to which the user is linked.  This is undesirable; we generally do not want an admin
+of `Customer A` to also have admin permissions for `Customer B` (to which the user may be linked as a mere ``learner``).
+Instead, the ``SystemWideEnterpriseUserRoleAssignment`` model should explicitly record *which* customer
+the assigned user has the given role in.
+
+Decisions
+=========
+
+Add an ``enterprise_customer`` field to the assignment model
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* It will be an optional field.  It should
+  also be a foreign key on ``EnterpriseCustomer.uuid``.
+* Can we ever make this a required field?  Specifically, will edX/internal users be problematic?
+* We will also add a boolean ``applies_to_all`` field, which is a wildcard that, if true,
+  indicates the use has the role assigned for every customer.  It should default to ``false``.
+* We will add a customer validator to enforce that the ``enterprise_customer`` field can be null
+  only when ``applies_to_all`` is true.
+
+We'll backfill the values of this field
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* We'll make the current, implicit assignments (that is, if a role is assigned to a user, that role
+  applies to the user in every customer to which the user is linked) explicit by filling
+  in the UUID value of the linked enterprise customer.
+* For ``enterprise_openedx_operator`` roles, the ``applies_to_all`` field will be marked true.
+
+Do a hand-audit of enterprise_admin role assignments
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* For every user who is effectively an admin of two different enterprise customers,
+  we'll have to manually determine (with help from our support team) which of the two customers
+  they are actually an admin of, and which they are simply a learner within (unless they're truly an admin of both).
```

Comment:
Since we have a foreign key on the enterprise uuid, how will this work if they *are* an admin of multiple enterprises? Will there be a row for each enterprise rather than each role/user?

Reply:
The applies_to_all flag just allows it to be null rather than contain multiple entries, so right now it looks like we allow you to be admin of 1 or all enterprises with no in between. That may be acceptable, just want to verify.

Reply:
Right, there would be one row for each (user, role, enterprise_customer) - except in the case of `applies_to_all=true`.  So if you were and enterprise of Pepsico and Coca-cola, there'd be rows like
```
('Talia', 'enterprise_admin', 'Pepsico'),
('Talia', 'enterprise_admin', 'Coca-cola'),
```

Reply:
Okay good. That makes sense.

Reply:
I guess another question is on a state change of going from "enterprise_admin" to the operator role, would the old rows be deleted or do we expect to have some logic that fetches by role and ignores the enterprise when looking at an operator role?

Reply:
Yeah, that logic already exists - if you have the operator role, the context/customer you have permissions for is `"*"`